### DISCUSSION TITLE
asyncua vs. opcua question

### DISCUSSION THREAD

**Discussion by rhb327** [FreeOpcUa/opcua-asyncio#454] [Q&A]
I'm working on an embedded Qt design that will interface with Emerson's DeltaV system via OPC UA.  On the embedded side I have a python OPC UA server based on either opcua or asyncua and I have a OPC UA client in the Qt user interface application.  I'm trying to select between these two libraries noting the following:

- asyncua seems to be the newer library and the primary path into the future (true?)
- opcua seems to have more development maturity to support OPC UA alarms, events and methods which I'm interested in supporting

Assuming the above are true, would you have a specific recommendation of which library would be best to use this design?

Thanks!
-Rich

**Comment by swamper123** [comment]
Hey ho,

- opcua-asyncio is the actual and primary repo. python-opcua is in maintainance mode (and probably will be for a long time)
- everything that python-opcua can, should be possible with asyncua as well (afaik we have a sync wrapper, which should deal with that. Haven't tested that, since I am mostly using the async way).  Asyncua has Events and methods already implemented (see the example folder). OPC UA Alarms are not implemented yet, but is in our [work in progress](https://github.com/FreeOpcUa/opcua-asyncio/projects/1) list.

Just a hint, FreeOpcUa has a [OPC-UA GUI client ](https://github.com/FreeOpcUa/opcua-client-gui)as well, containing PyQT5.

**Comment by oroulet** [comment]
and also if you want to do "real" embeded system, then python (and QT) is not the right language. you have an open source opcua in C project on github too.
